apollo-federation/src/query_plan/optimize.rs

```diff
@@ -0,0 +1,755 @@
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(true)
+    }
+
+    fn do_merge_with(&self, other: &FieldsConflictValidator) -> Result<bool, FederationError> {
+        for (response_name, self_fields) in self.by_response_name.iter() {
+            let Some(other_fields) = other.by_response_name.get(response_name) else {
+                continue;
+            };
+
+            // We're basically checking
+            // [FieldsInSetCanMerge](https://spec.graphql.org/draft/#FieldsInSetCanMerge()), but
+            // from 2 set of fields (`self_fields` and `other_fields`) of the same response that we
+            // know individually merge already.
+            for (self_field, self_validator) in self_fields {
+                for (other_field, other_validator) in other_fields {
+                    if !self_field.types_can_be_merged(other_field)? {
+                        return Ok(false);
+                    }
+
+                    let p1 = self_field.parent_type_position();
+                    let p2 = other_field.parent_type_position();
+                    if p1 == p2 || !p1.is_object_type() || !p2.is_object_type() {
+                        // Additional checks of `FieldsInSetCanMerge` when same parent type or one
+                        // isn't object
+                        if self_field.data().name() != other_field.data().name()
+                            || self_field.data().arguments != other_field.data().arguments
+                        {
+                            return Ok(false);
+                        }
+                        if let Some(self_validator) = self_validator {
+                            if let Some(other_validator) = other_validator {
+                                if !self_validator.do_merge_with(other_validator)? {
+                                    return Ok(false);
+                                }
+                            }
+                        }
+                    } else {
+                        // Otherwise, the sub-selection must pass
+                        // [SameResponseShape](https://spec.graphql.org/draft/#SameResponseShape()).
+                        if let Some(self_validator) = self_validator {
+                            if let Some(other_validator) = other_validator {
+                                if !self_validator.has_same_response_shape(other_validator)? {
+                                    return Ok(false);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(true)
+    }
+
+    fn do_merge_with_all<'a>(
+        &self,
+        mut iter: impl Iterator<Item = &'a FieldsConflictValidator>,
+    ) -> Result<bool, FederationError> {
+        iter.try_fold(true, |acc, v| Ok(acc && self.do_merge_with(v)?))
+    }
+}
```

Comment:
But, `is_ok_and(...)` would treat `FederationError` as `false`.
I used `try_fold`, so I can return `FederationError` to caller.

I looked for `try_all` or something like that. But, there were some discussions at Rust community on this. But, it seems not determined, yet. (https://github.com/rust-lang/rfcs/pull/3233)

Reply:
It would treat `FederationError` as `false`, yes. 

When I was reading this functions' usage, I thought the error was ignored entirely, which would make this function a bit redundant. I missed the `?` operand. Makes sense then!